File: DataStructure/5.Tree/Group1/seoyoon/1991.cpp
#include <iostream>
#include <map>

using namespace std;

struct Node {
    char left;
    char right;
};

map<char, Node> tree;

void preorder(char root) {
    if (root == '.') return;
    cout << root;
    preorder(tree[root].left);
    preorder(tree[root].right);
}

void inorder(char root) {
    if (root == '.') return;
    inorder(tree[root].left);
    cout << root;
    inorder(tree[root].right);
}

void postorder(char root) {
    if (root == '.') return;
    postorder(tree[root].left);
    postorder(tree[root].right);
    cout << root;
}

int main() {
    int n;
    cin >> n;

    for (int i = 0; i < n; i++) {
        char parent, left, right;
        cin >> parent >> left >> right;
        tree[parent] = { left, right };
    }

    preorder('A');
    cout << '\n';
    inorder('A');
    cout << '\n';
    postorder('A');
    cout << '\n';

    return 0;
}

File: DataStructure/5.Tree/Group3/dooyeon/1991.py
class BinaryTree:
    def __init__(self):
        self.tree = {}

    def add_node(self, parent, left, right):
        self.tree[parent] = (left, right)

    def preorder(self, node):
        if node == '.':
            return ''
        left, right = self.tree[node]
        return node + self.preorder(left) + self.preorder(right)

    def inorder(self, node):
        if node == '.':
            return ''
        left, right = self.tree[node]
        return self.inorder(left) + node + self.inorder(right)

    def postorder(self, node):
        if node == '.':
            return ''
        left, right = self.tree[node]
        return self.postorder(left) + self.postorder(right) + node

n = int(input())
binary_tree = BinaryTree()

for i in range(n):
    parent, left, right = input().split()
    binary_tree.add_node(parent, left, right)

root = 'A'
print(binary_tree.preorder(root))
print(binary_tree.inorder(root))
print(binary_tree.postorder(root))